### PR TITLE
add tokenize to node client api

### PR DESCRIPTION
**Pull request by jraymakers**
Add plumbing to expose the Tokenize method through the Node client API.

The signature of this method follows the [corresponding method from duckdb-wasm](https://github.com/duckdb/duckdb-wasm/blob/main/packages/duckdb-wasm/src/bindings/bindings_interface.ts#L14). It returns an object containing two arrays of numbers, instead of an array of objects, each with two number properties (as the underlying C++ API does), presumably to reduce the number of JS objects that need to be created.

I opted to call the Tokenize method synchronously, since it's lightweight and doesn't require any database resources; it's a static method of the parser. If there's a reason to move the call to the worker thread, let me know.

In addition to exposing the method itself, the `TokenType` enum is also exposed. This requires some code on the C++ side in order to add the runtime enum object to the NodeDuckDB Addon, in addition to the type definition on the TypeScript side.

A basic unit test suite is included.

### CODE REVIEW

File: tools/nodejs/src/duckdb_node.cpp [hunk 1]
```diff
@@ -12,15 +12,26 @@ NodeDuckDB::NodeDuckDB(Napi::Env env, Napi::Object exports) {
 	statement_constructor = node_duckdb::Statement::Init(env, exports);
 	query_result_constructor = node_duckdb::QueryResult::Init(env, exports);
 
-	exports.DefineProperties({
-	    DEFINE_CONSTANT_INTEGER(exports, node_duckdb::Database::DUCKDB_NODEJS_ERROR, ERROR) DEFINE_CONSTANT_INTEGER(
+	auto token_type_enum = Napi::Object::New(env);
+	token_type_enum.Set("IDENTIFIER", 0);
```
Comment (Mause) [inline tools/nodejs/src/duckdb_node.cpp:16]: We should probably also expose the inverted mapping as well, to be a "true" typescript enum

File: tools/nodejs/src/database.cpp [hunk 2]
```diff
@@ -364,4 +366,31 @@ Napi::Value Database::RegisterReplacementScan(const Napi::CallbackInfo &info) {
...
+Napi::Value Database::Tokenize(const Napi::CallbackInfo &info) {
+	auto env = info.Env();
+
+	if (info.Length() < 1) {
+		throw Napi::TypeError::New(env, "Text argument expected");
+	}
+
+	std::string text = info[0].As<Napi::String>();
+
+	auto tokens = duckdb::Parser::Tokenize(text);
```
Comment (Mause) [inline tools/nodejs/src/database.cpp:378]: How well does this scale on huge queries?

File: tools/nodejs/src/database.cpp [hunk 3]
```diff
@@ -364,4 +366,31 @@ Napi::Value Database::RegisterReplacementScan(const Napi::CallbackInfo &info) {
...
+	auto numTokens = tokens.size();
+
+	auto offsets = Napi::Array::New(env, numTokens);
+	auto types = Napi::Array::New(env, numTokens);
+
+	for (size_t i = 0; i < numTokens; i++) {
+		auto token = tokens[i];
+		offsets.Set(i, token.start);
```
Comment (Mause) [inline tools/nodejs/src/database.cpp:386]: Not exposing the token lengths? I'm assuming due to whitespace we couldn't calculate them from purely the offsets

File: tools/nodejs/test/tokenize.test.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import * as assert from 'assert';
+import * as duckdb from '..';
+
+describe('tokenize', function () {
```
Comment (Mause) [inline tools/nodejs/test/tokenize.test.ts:4]: Can you add a test for handing invalid syntax?

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/nodejs/src/database.cpp (hunk 3 above)
Comment (jraymakers) [reply tools/nodejs/src/database.cpp:386]: It's a good point, but the underlying DuckDB Tokenize method only exposes `start` and `type`. I'm just exposing it as-is.

File: tools/nodejs/src/database.cpp (hunk 2 above)
Comment (jraymakers) [reply tools/nodejs/src/database.cpp:378]: This just runs the scanner, not the full parser, so I assume the runtime is pretty close to linear in the input string.

File: tools/nodejs/test/tokenize.test.ts (hunk 4 above)
Comment (jraymakers) [reply tools/nodejs/test/tokenize.test.ts:4]: Ah, good point, will do.

File: tools/nodejs/src/duckdb_node.cpp (hunk 1 above)
Comment (jraymakers) [reply tools/nodejs/src/duckdb_node.cpp:16]: Hm, yes, I'll look into that.

Comment (jraymakers) [reply tools/nodejs/src/duckdb_node.cpp:17]: Done.

File: tools/nodejs/test/tokenize.test.ts (hunk 4 above)
Comment (jraymakers) [reply tools/nodejs/test/tokenize.test.ts:4]: I added one test, but of course there are many variations on invalid syntax. The scanner doesn't throw on invalid syntax, though - it'll just mark tokens as different types.

File: tools/nodejs/src/database.cpp (hunk 2 above)
Comment (jraymakers) [reply tools/nodejs/src/database.cpp:378]: The underlying implementation is [here](https://github.com/duckdb/duckdb/blob/main/third_party/libpg_query/src_backend_parser_parser.cpp#L99).

**Comment by Mytherin** [comment]
Thanks for the PR! We are in the process of moving the NodeJS package to its own repository - see [here](https://github.com/duckdb/duckdb-node). Could you make the contribution there instead?

**Comment by jraymakers** [comment]
> Thanks for the PR! We are in the process of moving the NodeJS package to its own repository - see [here](https://github.com/duckdb/duckdb-node). Could you make the contribution there instead?

Sure, I'll move this PR over to that repo.

**Comment by Mytherin** [comment]
Node has now been fully removed from this repo

**Pull request closed without merging**